vendor/parser-jane/for-parser-standard/parser.mly

```diff
@@ -153,9 +153,31 @@ let mk_attr ~loc name payload =
   Attr.mk ~loc name payload
 
 let mkpat_with_modes ~loc ~pat ~cty ~modes =
-  match cty, modes with
-  | None, [] -> pat
-  | cty, modes -> mkpat ~loc (Ppat_constraint (pat, cty, modes))
+  match pat.ppat_desc with
+  | Ppat_constraint (pat', cty', modes') ->
+    begin match cty, cty' with
+    | Some _, None ->
+      { pat with
+        ppat_desc = Ppat_constraint (pat', cty, modes @ modes');
+        ppat_loc = make_loc loc
+      }
+    | None, _ ->
+      { pat with
+        ppat_desc = Ppat_constraint (pat', cty', modes @ modes');
+        ppat_loc = make_loc loc
+      }
+    | _ ->
+      mkpat ~loc (Ppat_constraint (pat, cty, modes))
+    end
+  | _ ->
+    begin match cty, modes with
+    | None, [] -> pat
```

Comment:
Can this case ever happen?  If no, I think either dropping this match or making this case `assert false` would be clearer.

Reply:
It can happen. There are a few places where `mkpat_with_modes` gets called with no modes and an optional core type